[dynamo] save/restore system random state more carefully

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145435

Fixes https://github.com/pytorch/pytorch/issues/145329.

We we need to save/restore system `random` state in 2 places:
- in `eval_frame.py`, we need to make sure that wrapper code between the user call to a `torch.compile`d function and the actual function call (intercepted by eval_frame.c) doesn't modify random state (https://github.com/pytorch/pytorch/blob/b2c89bc115123aea8e075e882ee121537ec92f89/torch/_dynamo/eval_frame.py#L532)
- in `eval_frame.c`, we need to make sure that guard eval and calling convert_frame don't modify random state (https://github.com/pytorch/pytorch/blob/b2c89bc115123aea8e075e882ee121537ec92f89/torch/csrc/dynamo/eval_frame.c#L575)

Followup - perhaps more global state from `convert_frame.py:preserve_global_state` can be moved to `eval_frame.py/c`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Differential Revision: [D68532640](https://our.internmc.facebook.com/intern/diff/D68532640)

williamwen42 committed:
[dynamo] restore system random state more carefully

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] restore system random state more carefully"


Fixes https://github.com/pytorch/pytorch/issues/145329.

How system `random` is restored:
- in eval_frame, we reset `random` state in a way that only random calls in the real call to the function in `_TorchDynamoContext.__call__._fn` modify `random` state.
- in convert_frame, we reset `random` state to before the start of tracing, at the C level (e.g. since logging calls that burn RNG can be made before `preserve_global_state`)

We use a separate random object `dynamo_random` in order to generate example random values during tracing without interference from possible rogue RNG burns (not yet tested).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
@williamwen42 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68532640).

williamwen42 committed:
Update on "[dynamo] restore system random state more carefully"


Fixes https://github.com/pytorch/pytorch/issues/145329.

How system `random` is restored:
- in eval_frame, we reset `random` state in a way that only random calls in the real call to the function in `_TorchDynamoContext.__call__._fn` modify `random` state.
- in convert_frame, we reset `random` state to before the start of tracing, at the C level (e.g. since logging calls that burn RNG can be made before `preserve_global_state`)

We use a separate random object `dynamo_random` in order to generate example random values during tracing without interference from possible rogue RNG burns (not yet tested).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68532640](https://our.internmc.facebook.com/intern/diff/D68532640)

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -293,8 +294,34 @@ static PyObject* dynamo_call_callback(
+
+  // store error if there was one
+#if IS_PYTHON_3_12_PLUS
+  PyObject* exc = PyErr_GetRaisedException();
+#else
+  PyObject *exc_type, *exc_value, *exc_traceback;
```
Are these objects owned an needing recounting?

williamwen42 replied:
No need, PyErr_SetRaisedException/PyErr_Restore steals the reference.

jansel commented on torch/csrc/dynamo/utils.cpp:
```diff
@@ -2,6 +2,43 @@
+// random utilities for C dynamo
+
+// random module reference
+py::object _random = py::module::import("random");
+
+// random.Random()
+// returns new reference
+PyObject* new_random_object() {
+  py::object rng = py::module::import("random").attr("Random")();
+  return rng.release().ptr();
+}
+
+// random
+// returns borrowed reference
+PyObject* random_module() {
+  return _random.ptr();
+}
+
+// rng.getstate()
+// rng can be random module or random.Random object
+// rng: borrowed reference
+// returns new reference
+PyObject* random_getstate(PyObject* rng) {
+  py::handle rng_h(rng);
+  py::object state = rng_h.attr("getstate")();
+  return state.release().ptr();
+}
+
+// rng.setstate(state)
+// rng can be random module or random.Random object
+// rng, state: borrowed references
+// no return value
+void random_setstate(PyObject* rng, PyObject* state) {
+  py::handle rng_h(rng), state_h(state);
+  rng_h.attr("setstate")(state_h);
+}
```
Is there no C++ API for this stuff?

williamwen42 replied:
Not that I'm aware of.

jansel commented on torch/csrc/dynamo/utils.h:
```diff
@@ -14,5 +22,21 @@
+
+#endif // __cplusplus
+
+PyObject* new_random_object();
+PyObject* random_module();
+PyObject* random_getstate(PyObject* rng);
+void random_setstate(PyObject* rng, PyObject* state);
```
Coment on the ownership model here?  Are these owned pointers being returned?

williamwen42 replied:
Moved ownership comments from .cpp file to .h file.

jansel reviewed (changes requested):
I know there was some existing code to save/restore the random state -- can that code be removed now?  Why do we need both?

williamwen42 commented:
@jansel the previous code to save/restore global `random` state was removed from convert_frame.py's `preserve_global_state`: https://github.com/pytorch/pytorch/pull/145435/files#diff-d20e797a5fba9e8dcfa019472f684862a7f616fd654a6600bc0a83795ea49fa1L220

williamwen42 committed:
Update on "[dynamo] restore system random state more carefully"


Fixes https://github.com/pytorch/pytorch/issues/145329.

How system `random` is restored:
- in eval_frame, we reset `random` state in a way that only random calls in the real call to the function in `_TorchDynamoContext.__call__._fn` modify `random` state.
- in convert_frame, we reset `random` state to before the start of tracing, at the C level (e.g. since logging calls that burn RNG can be made before `preserve_global_state`)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68532640](https://our.internmc.facebook.com/intern/diff/D68532640)

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] restore system random state more carefully"


Fixes https://github.com/pytorch/pytorch/issues/145329.

How system `random` is restored:
- in eval_frame, we reset `random` state in a way that only random calls in the real call to the function in `_TorchDynamoContext.__call__._fn` modify `random` state.
- in convert_frame, we reset `random` state to before the start of tracing, at the C level (e.g. since logging calls that burn RNG can be made before `preserve_global_state`)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68532640](https://our.internmc.facebook.com/intern/diff/D68532640)

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] restore system random state more carefully"


Fixes https://github.com/pytorch/pytorch/issues/145329.

How system `random` is restored:
- in eval_frame, we reset `random` state in a way that only random calls in the real call to the function in `_TorchDynamoContext.__call__._fn` modify `random` state.
- in convert_frame, we reset `random` state to before the start of tracing, at the C level (e.g. since logging calls that burn RNG can be made before `preserve_global_state`)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68532640](https://our.internmc.facebook.com/intern/diff/D68532640)

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] save/restore system random state more carefully"


Fixes https://github.com/pytorch/pytorch/issues/145329.

We we need to save/restore system `random` state in 2 places:
- in `eval_frame.py`, we need to make sure that wrapper code between the user call to a `torch.compile`d function and the actual function call (intercepted by eval_frame.c) doesn't modify random state (https://github.com/pytorch/pytorch/blob/b2c89bc115123aea8e075e882ee121537ec92f89/torch/_dynamo/eval_frame.py#L532)
- in `eval_frame.c`, we need to make sure that guard eval and calling convert_frame don't modify random state (https://github.com/pytorch/pytorch/blob/b2c89bc115123aea8e075e882ee121537ec92f89/torch/csrc/dynamo/eval_frame.c#L575)

Followup - perhaps more global state from `convert_frame.py:preserve_global_state` can be moved to `eval_frame.py/c`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68532640](https://our.internmc.facebook.com/intern/diff/D68532640)

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] save/restore system random state more carefully"


Fixes https://github.com/pytorch/pytorch/issues/145329.

We we need to save/restore system `random` state in 2 places:
- in `eval_frame.py`, we need to make sure that wrapper code between the user call to a `torch.compile`d function and the actual function call (intercepted by eval_frame.c) doesn't modify random state (https://github.com/pytorch/pytorch/blob/b2c89bc115123aea8e075e882ee121537ec92f89/torch/_dynamo/eval_frame.py#L532)
- in `eval_frame.c`, we need to make sure that guard eval and calling convert_frame don't modify random state (https://github.com/pytorch/pytorch/blob/b2c89bc115123aea8e075e882ee121537ec92f89/torch/csrc/dynamo/eval_frame.c#L575)

Followup - perhaps more global state from `convert_frame.py:preserve_global_state` can be moved to `eval_frame.py/c`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68532640](https://our.internmc.facebook.com/intern/diff/D68532640)

[ghstack-poisoned]

williamwen42 commented:
@williamwen42 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68532640).

williamwen42 commented:
@pytorchbot merge

williamwen42 commented:
@williamwen42 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68532640).

williamwen42 commented:
@pytorchbot merge

williamwen42 commented:
Closing as there is a reattempt PR
